pypose/module/cameras.py

```diff
@@ -0,0 +1,150 @@
+        Returns:
+            points (torch.Tensor): A tensor of shape (B, N, 2) where B is the batch size.
+        """
+        raise NotImplementedError
+
+    def reprojection_error(self, points, pixels):
+        """
+        Calculate the reprojection error.
+        Args:
+            points: The object points in world coordinate. The shape is (..., N, 3).
+            pixels: The image points. The shape is (..., N, 2).
+        Returns:
+            error: The reprojection error. The shape is (..., ).
+        """
+        # Calculate the image points
+        img_repj = self.world2pixel(points)
+
+        error = torch.linalg.norm(img_repj - pixels, dim=-1)
+        error = error.mean(dim=-1)
+
+        return error
+
+
+class PerspectiveCameras(CamerasBase):
+    r"""
+    A class which stores a batch of parameters to generate a batch of
+    transformation matrices using the multi-view geometry convention for
+    perspective camera.
+
+    Args:
+        pose (LieTensor): A tensor of shape (B, *) where B is the batch size.
+        intrinsics (Optional): A projection matrix of shape (N, 3, 3)
+
+    Examples:
+        >>> import torch
+        >>> import pypose
+        >>> # create some random data
+        >>> pose = pypose.SE3([ 0.0000, -8.0000,  0.0000,  0.0000, -0.3827,  0.0000,  0.9239])
+        >>> f = 2
+        >>> img_size = (7, 7)
+        >>> projection_matrix = torch.tensor([[f, 0, img_size[0] / 2,], [0, f, img_size[1] / 2,], [0, 0, 1, ]])
+        >>> pts_w = torch.tensor([[ 2.8284,  8.0000,  0.0000], [ 2.1213,  8.0000,  0.7071], [ 0.7071,  9.0000,  0.7071], [ 0.7071,  8.0000,  0.7071], [ 5.6569, 13.0000, -1.4142]])
+        >>> # instantiate the camera
+        >>> camera = pypose.module.PerspectiveCameras(pose=pose, intrinsics=projection_matrix)
+        >>> # transform the points to image coordinates
+        >>> img_pts = camera.world2pixel(pts_w)
+        >>> img_pts
+        tensor([[5.4998, 3.5000],
+                [4.4999, 3.5000],
+                [3.4999, 5.5000],
+                [3.4999, 3.5000],
+                [6.8329, 6.8330]])
+    """
+
+    def __init__(
+            self,
+            pose: Optional[pypose.LieTensor] = None,
+            intrinsics: Optional[torch.Tensor] = None,
+    ):
+        super().__init__(pose=pose)
+        if intrinsics is not None:
+            self.intrinsics = intrinsics
+        elif self.intrinsics is None:
+            self.intrinsics = torch.eye(3)
+
+    def camera2pixel(self, points):
+        r"""
+        Transform camera view coordinate points to pixel locations.
+        Args:
+            points (torch.Tensor): A tensor of shape (B, N, 3) where B is the batch size.
+        Returns:
+            points (torch.Tensor): A tensor of shape (B, N, 2) where B is the batch size.
+        """
+        img_repj = points.matmul(self.intrinsics.mT)
+        return img_repj[..., :2] / img_repj[..., 2:]
+
+    def is_perspective(self):
```

Comment:
Design: is this function really necessary? Is there a camera class other than PerspectiveCamera that is also perspective? If not, I would suggest removing this function (or at least implement it in the base class, and default to return False)

Reply:
not really necessary. going to remove it